### DISCUSSION TITLE
Reverse proxy

### DISCUSSION THREAD

**Discussion by Lumycea** [clicepfl/website#14] [Ideas]
We need to investigate which reverse proxy to use. Ideas:
- [nginx](https://www.nginx.com/)
- [Traefik](https://doc.traefik.io/traefik/)
- [Caddy](https://caddyserver.com/)

**Comment by Voxanimus** [comment]
I read some stuff about the 3 reverse proxy. It seems that caddy sounds to have very simple config file and handle automatically the certifications. As it's more like a server if we need to extend add functionality we won't had dependencies to other tools whereas traefik seems more inflexible.

**Reply by Lumycea** [reply]
The advantage of Traefik is that it can dynamically resolve some endpoints, and does not require any config updates, reloads or restarts. \
For example, if you add a Nextcloud service at https://clic.epfl.ch/nextcloud, and add the correct tags to the __service__, Traefik will automatically detect it and redirect the corresponding request to it. Caddy would need a config update, either through a (partial) reupload of the config, or using the API endpoints.

> As it's more like a server if we need to extend add functionality we won't had dependencies to other tools whereas traefik seems more inflexible

I did not understand. Do you mean that Caddy will never need any additional process to fulfill its tasks ? If so, I don't see why Traefik cannot do it.

**Reply by Voxanimus** [reply]
Yeah, I didn't though that we host several services when I read those stuff. From this point of view it seems for legit to use traeffik.
For the second part, Traeffik to be only a pure reverse proxy while Caddy, as far as I understood, is more volatile. It can be used as a file server for example. Maybe it is completely overkill, but I think that if people in the future want to add completely different services in the infrastructure it could be a nice idea. Note that everything I say is very approximative since I never touch this tools.

**Reply by Lumycea** [reply]
To my understanding, the current idea for the infrastructure is to have everything in containers, and manage them through an orchestration tool. I think it would break that to serve something through another mean, like static content through Caddy's webserver API. If we were to do that, I think we should better use a container which serves the content, and the reverse proxy would simply redirect request to the container.

Keeping the scheme for the infrastructure is, in my opinion, very important to avoid the server to be more and more unmaintainable in the future, as we would add services here and there.

Everything I say is purely subjective, and at the moment I just got a look at Traefik and Caddy, and there are many other solutions. I'll dive deeper soon. 

It would be great to have other points of view too

**Reply by Lumycea** [reply]
So I've been experimenting with Traefik, and here are my conclusions:
Pros:
- The use of labels on container is very dynamic
- The whole tool seems modular and powerful and could be great for future project on the server

Cons:
- The learning curve is very steep, I got a really hard time learning to setup even the most basic configuration, and I'm still stuck on simple problems
- I found the documentation hard to understand and it was not easy to find what I wanted
- Understanding the issues is not easy, as the logging (even in debug mode) is sparse
- It's not possible to reload the static configuration while the container is running, and restarting a docker swarm service is not something you should do

Overall, I think Traefik may be a bit too much for this server. We would loose much more time learning to use it and debug it than we would need to learn and setup other reverse proxies, like nginx.

I'm gonna have a look to Caddy this week-end

**Reply by mochiruntime** [reply]
Thank you for your research and valuable feedback! I agree, if the knowledge cost of Traefik is too high it's probably not worth it for our use case. I'm also curious about Caddy, since they market it as a very simple, low-config, production-ready reverse proxy. I'll be happy to read your opinion on it :+1:

**Reply by Lumycea** [reply]
I've tried a bit Caddy, and I'm not very happy. I've been struggling for the past 45 just trying to access the API when Caddy is in a container, and it's still not working. 

The documentation sometimes has instructions about legacy behavior from Caddy 1, which makes it very annoying when you try to fix something and after 15 minutes you discover that environments variables are not used anymore (but the caddy executable still has an option to show them)

Furthermore, Caddy does not seem to be used in docker, a lot of things are done through the `caddy` command, thus requiring to be inside the docker (losing any other tools like vim or curl)

For now, I would suggest we use Nginx, as we are several to already have some experience with it. I will try to deploy it from a container on my setup. I might also give another try to Caddy in the meantime, if I have the time

**Reply by Lumycea** [reply]
So I've tried Nginx inside a container, and there are no noticeable differences from the classic setup. I think we can go for this one

**Reply by mochiruntime** [reply]
Thank you for your work and your time! Sounds good to me, we can keep nginx as our reverse proxy for now.

**Comment by mochiruntime** [comment]
Bumping this thread, I've been using Caddy on my home lab and I really like its simplicity and sane defaults. In my opinion it would fit well with the "pit of success" approach we're trying to have in this project, I think it's more foolproof than e.g. nginx. For instance a Caddyfile can be as simple as:

```
clic.epfl.ch {
     reverse_proxy <container_name>:<port>
}
```

and SSL would be handled automatically by Let's Encrypt. Adding a certificate file is super easy as well:

```
clic.epfl.ch {
    tls cert.pem key.pem
    reverse_proxy <container_name>:<port>
}
```

I will add a working example to the repo in a PR
